### PR TITLE
feat: v7

### PR DESCRIPTION
**Pull request by dannyhw**
Issue:

## What I did

started work on v7

## How to test

run the example app
`yarn workspace expo-example start`

# TODO

- [x] update all storybook dependencies
- [x] core functionality working
- [x] addons
- [x] fix storiesOf
- [x] fix server
  - server needs a way to set globals from the server to the browser
- [x] fix compat with 7.6 alpha7 + 
  - need to fix @storybook/global not found
- [x] fix actions causing an error due to no crypto get random

# documentation todo

- [x] update readmes
- [x] update starters
- [x] update template
- [x] migration guide
- [x] stories of compatibility mode migration steps
- [x] update tutorial

### CODE REVIEW

**Comment by dannyhw** [comment]
@SocketSecurity ignore-all

**Comment by thomazcapra** [comment]
👀

File: packages/react-native/V6.js [hunk 1]
```diff
@@ -0,0 +1 @@
+export * from './dist/V6';
```
Comment (mljlynch) [inline packages/react-native/V6.js:1]: question: why is this file necessary?

Comment (dannyhw) [reply packages/react-native/V6.js:1]: Its for backwards compatibility with the v6 store. In v7 there is a new story store which changes the way that we load stories and simplifies the setup.

Comment (dannyhw) [reply packages/react-native/V6.js:1]: For example storiesOf requires the v6 store and isn't compatible with the v7 store

**Comment by YOEL311** [comment]
Hey 
how i can use your branch when I create a new React native storyboard app with a template

``npx react-native init MyApp --template react-native-template-storybook``

**Comment by dannyhw** [comment]
@YOEL311 I haven't added it to the template but you could follow these steps to get it working

Update react native storybook dependencies to v7.0.0-alpha.8
 
```
"@storybook/react-native": "^7.0.0-alpha.8",
"@storybook/addon-ondevice-actions": "^7.0.0-alpha.8",
"@storybook/addon-ondevice-backgrounds": "^7.0.0-alpha.8",
"@storybook/addon-ondevice-controls": "^7.0.0-alpha.8",
```

change .storybook/index to be

```
import { view } from './storybook.requires';

const StorybookUIRoot = view.getStorybookUI({});

export default StorybookUIRoot;
```

Regenerate the storybook.requires file with update-stories or storybook-generate depending on which script you have in your package json (the one that runs sb-rn-get-stories).

Also update metro config to include

```
transformer: {
  unstable_allowRequireContext: true,
},
```

Make sure to also install 7.6.0-alpha.4 or newer for other storybook packages not directly react-native related. For example if you have storybook/react then change it to `"@storybook/react": "7.6.0-alpha.4",`

It's also currently necessary to install the package called `assert` though that will be removed soon.

**Comment by YOEL311** [comment]
@dannyhw 
Thank you for your response
Should it work without any issues?

**Comment by dannyhw** [comment]
@YOEL311 well its an early version so I can't promise no changes and no bugs but all features are implemented and working based on my own testing. Still need to finalise types and the server, but those shouldn't block usage.

**Comment by simonadenic** [comment]
Hey @dannyhw, is there any update on when this is going to be released?

**Comment by dannyhw** [comment]
@simonadenic I've been on holiday the last month thats pushed things back a bit but the alpha is mostly working already. Depends how much time I have in the following weeks.

**Comment by dannyhw** [comment]
if anyone is trying the alpha and getting the storybook/global is not found error then add this to metro config

```
resolveRequest: (context, moduleName, platform) => {
        if (moduleName === '@storybook/addon-actions') {
          const defaultResolveResult = context.resolveRequest(context, moduleName, platform);
          defaultResolveResult.filePath = defaultResolveResult.filePath.replace(
            'index.js',
            'index.mjs'
          );
          return defaultResolveResult;
        }
        return context.resolveRequest(context, moduleName, platform);
},
```

___ 

Edit this was resolved in 7.6.8 of addon-actions

**Comment by dannyhw** [comment]
just published the 7.6.7-alpha.1 version for compatibility with the latest storybook version. You will need to apply the above metro config if you are using addon actions.

**Comment by dannyhw** [comment]
the server package is no longer possible to update due to changes to storybook, instead to replace it for v7 I've published an addon that runs inside the web storybook and has the same functionality.

**Comment by dannyhw** [comment]
https://github.com/storybookjs/react-native/assets/3481514/73e264b5-016c-4e76-ace7-2abd47072212

**Comment by dannyhw** [comment]
7.6.8-alpha.0 is out with latest fixes.

**Comment by dannyhw** [comment]
7.6.9-alpha.0 out with more fixes and updated types. We will likely stick with 7.6.9 as the release version now that all the necessary upstream fixes are in 7.6.9.

**Comment by stevoland** [comment]
Hi @dannyhw, fantastic stuff, tried out 7.6.8-alpha.0 on my project and it's looking great, addon-react-native-server is an unexpected bonus.

Saw one issue so far, repro in [this commit](https://github.com/stevoland/react-native/commit/a76f075371b28e266dc4a21a2d35b82f99f2b7a0) - multiple stories config options using titlePrefix cause all stories to be grouped under the first folder

**Comment by dannyhw** [comment]
@stevoland thanks for testing it out and especially for sending the commit so I can debug🙏 

I will check that tomorrow morning.

**Comment by dannyhw** [comment]
types for `@storybook/react-native` are being removed and user should use the types from `@storybook/react` instead, this is to increase compatibility and reduce duplication. Since `@storybook/react-native` now depends on `@storybook/react` for a lot more code re-use the package will already be included.

for example 
```
import type { Meta, StoryObj } from '@storybook/react';

const BooleanExample: Meta<typeof Switch> = {
  title: 'ControlExamples/Boolean Control',
  component: Switch,
};

export default BooleanExample;

type BooleanStory = StoryObj<typeof Switch>;

export const Basic: BooleanStory = {};
```

This way stories should look mostly identical for react and react-native

**Comment by dannyhw** [comment]
also I pushed another upstream fix to remove the acorn parser warning, so we'll go with 7.6.10 now 😅 . I want that to be the version we release on. This will also make 7.6.10 the minimum required version for core/web storybook packages like `@storybook/react`

**Comment by dannyhw** [comment]
version v7.6.10-alpha.0 contains the latest fix and the types removal

**Comment by awinograd** [comment]
I _think_ i've successfully upgraded to v7.6.10-alpha.0. All of our stories appear to be rendering as expected without too much issue. The only thing I had to change that wasn't in the updated README was adding `mjs` to sourceExts in metro config. Some additional details in discord:

https://discord.com/channels/486522875931656193/490822534233849873/1197984765979140117

EDIT: Forgot to say thank you @dannyhw for maintaining such a useful part of the storybook ecosystem

**Comment by dannyhw** [comment]
@awinograd thanks :)

That mjs in sourceExts is now automatically added in expo 50 so I guess thats why I don't see that issue. I can add it to the migration steps in case.

**Comment by dannyhw** [comment]
added https://github.com/storybookjs/react-native/pull/528 to this pr thanks to @patlux for your contribution 🙏 

This means you can now use mmkv or any other storage solution instead of asyncStorage

You will now need to now specify your storage config like:

```
import AsyncStorage from '@react-native-async-storage/async-storage';
const StorybookUIRoot = view.getStorybookUI({
  storage: {
    getItem: AsyncStorage.getItem,
    setItem: AsyncStorage.setItem,
  },
})
```

**Comment by dannyhw** [comment]
You can use this to try out v7

```
npx create-expo-app --template expo-template-storybook@next AwesomeStorybook
```

https://github.com/storybookjs/react-native/assets/3481514/9c5de428-f58f-4f05-b30f-7fdfd848927e

**Comment by dannyhw** [comment]
Hey everyone we've release 7.6.10-rc.0 as our release candidate, please give it a test 🙏
